mixer/pkg/config/registry/registry.go

```diff
@@ -0,0 +1,208 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"fmt"
+
+	"encoding/base64"
+	"io/ioutil"
+	"regexp"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/hashicorp/go-multierror"
+
+	adapter "istio.io/api/mixer/adapter/model/v1beta1"
+	tmpl "istio.io/api/mixer/adapter/model/v1beta1"
+	"istio.io/istio/pkg/log"
+)
+
+// TemplateMetadata contains info about a template
+type TemplateMetadata struct {
+	Name          string
+	FileDescSet   *descriptor.FileDescriptorSet
+	FileDescProto *descriptor.FileDescriptorProto
+}
+
+// AdapterMetadata contains info about an adapter
+type AdapterMetadata struct {
+	Info               *adapter.Info
+	SupportedTemplates []string
+}
+
+// Registry to find metadata about templates and adapters
+type Registry interface {
+	GetAdapter(name string) (adapter *AdapterMetadata, found bool)
+	GetTemplate(name string) (template *TemplateMetadata, found bool)
+}
+
+// registry to ingest templates and adapters. It is single threaded.
+type registry struct {
+	adapters  map[string]*AdapterMetadata
+	templates map[string]*TemplateMetadata
+}
+
+// New creates a `Registry` from given adapter infos.
+// Note: For adding built-in templates that are not associated with any adapters, supply the `Info` object with
+// only `templates`, leaving other fields to default empty.
+func New(infos []adapter.Info) (Registry, error) {
+	r := &registry{make(map[string]*AdapterMetadata), make(map[string]*TemplateMetadata)}
+	var resultErr error = nil
+	log.Debugf("registering %#v", infos)
+
+	for _, info := range infos {
+		if old, ok := r.GetAdapter(info.Name); ok {
+			// duplicate entry found
+			resultErr = multierror.Append(resultErr,
+				fmt.Errorf("duplicate registration for adapter '%s' : new = %v old = %v", info.Name, info, old))
+			continue
+		}
+
+		tmplNames, err := r.ingestTemplates(info.Templates)
+		if err != nil {
+			resultErr = multierror.Append(resultErr, err)
+			continue
+		}
+
+		// empty adapter name means just the template needs to be ingested.
+		if len(info.Name) != 0 {
+			r.adapters[info.Name] = &AdapterMetadata{Info: &info, SupportedTemplates: tmplNames}
+		}
+	}
+
+	if resultErr != nil {
+		log.Error(resultErr.Error())
+	}
+
+	return r, resultErr
+}
+
+// GetAdapterInfo returns a AdapterMetadata for a adapter with the given name.
+func (r *registry) GetAdapter(name string) (b *AdapterMetadata, found bool) {
```

Comment:
I'm not sure what Go conventions say about this, but since b is a pointer type, you don't need to return a pointer and a bool. Just return the pointer, or nil if not found.

@douglas-reid I'm sure you know what Go conventions are for this pattern...